In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
path = os.path.normpath(os.getcwd() + os.sep + os.pardir) + "\[1] Scraping Data" + "\cars.csv"
print(path)

H:\Projects\Epsilon InternShip\Cars Price Prediction\[1] Scraping Data\cars.csv


In [4]:
data = pd.read_csv(path ,encoding='latin-1')
data.head()

Make       Model   CC  Used since          Km Transmission       City  \
0    Skoda       Rapid  NaN      2014.0   89,000 Km    automatic  Nasr city   
1    Skoda  Octavia A7  NaN      2015.0  120,000 Km    automatic      Cairo   
2  Renault     Sandero  NaN      2016.0   70,000 Km    automatic   El Haram   
3       MG          ZS  NaN      2020.0   21,000 Km    automatic   El Haram   
4       MG           5  NaN      2021.0   28,000 Km    automatic   El Haram   

   Color Body Style Fuel  ...  Alloy wheels  Rear camera    GPS  Fog light  \
0  Black        NaN  gas  ...         False        False  False       True   
1   Blue        NaN  gas  ...          True         True   True       True   
2    Red        NaN  gas  ...         False        False  False      False   
3    Red        NaN  gas  ...          True        False   True       True   
4    Red        NaN  gas  ...          True        False  False       True   

   Rear spoiler  Cruise control  Power Seats  Central lock  Alarm        Price  
0         False           False         True          True   True  225,000 EGP  
1          True            True         True          True   True  295,000 EGP  
2         False           False        False         False  False  200,000 EGP  
3         False            True         True          True   True  360,000 EGP  
4         False           False         True         False   True  340,000 EGP  

[5 rows x 51 columns]

In [5]:
print(f'There are {data.shape[0]} rows and {data.shape[1]} columns')

There are 38611 rows and 51 columns


In [6]:
data['Km'].value_counts()

0 Km          7772
200,000 Km    2485
150,000 Km    1400
100,000 Km     950
130,000 Km     816
              ... 
408,000 Km       1
21,500 Km        1
130,500 Km       1
57,310 Km        1
186,100 Km       1
Name: Km, Length: 1032, dtype: int64

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38611 entries, 0 to 38610
Data columns (total 51 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Make                        37745 non-null  object 
 1   Model                       37745 non-null  object 
 2   CC                          19067 non-null  object 
 3   Used since                  37745 non-null  float64
 4   Km                          37745 non-null  object 
 5   Transmission                37745 non-null  object 
 6   City                        37745 non-null  object 
 7   Color                       37745 non-null  object 
 8   Body Style                  21285 non-null  object 
 9   Fuel                        37745 non-null  object 
 10  Leather seats               38611 non-null  bool   
 11  Closing mirrors             38611 non-null  bool   
 12  Intelligent parking system  38611 non-null  bool   
 13  Sunroof                     386

there are many null values in our dataset , we have 40 Boolean Column and 10 objects

In [8]:
data.isnull().sum()

Make                            866
Model                           866
CC                            19544
Used since                      866
Km                              866
Transmission                    866
City                            866
Color                           866
Body Style                    17326
Fuel                            866
Leather seats                     0
Closing mirrors                   0
Intelligent parking system        0
Sunroof                           0
Fabric brushes                    0
Air Conditioning                  0
Power Steering                    0
Remote Keyless                    0
Electric mirrors                  0
Front Power Windows               0
Back Power Windows                0
Tinted Glass                      0
CD Player                         0
DVD Player                        0
Bluetooth                         0
Multifunction                     0
Cassette Radio                    0
AUX                         

we have 866 null values in price Columns , let's drop it

In [9]:
data = data[data['Price'].notna()]

In [10]:
# checking 
data.isnull().sum()

Make                              0
Model                             0
CC                            18678
Used since                        0
Km                                0
Transmission                      0
City                              0
Color                             0
Body Style                    16460
Fuel                              0
Leather seats                     0
Closing mirrors                   0
Intelligent parking system        0
Sunroof                           0
Fabric brushes                    0
Air Conditioning                  0
Power Steering                    0
Remote Keyless                    0
Electric mirrors                  0
Front Power Windows               0
Back Power Windows                0
Tinted Glass                      0
CD Player                         0
DVD Player                        0
Bluetooth                         0
Multifunction                     0
Cassette Radio                    0
AUX                         

In [11]:
def show_column_details(col):
    global data
    print(f'for feature {col} and it type is {data[col].dtype}')
    print(f'Number of Nulls is {data[col].isna().sum()}')
    print(f'Number of Unique Values is {len(data[col].unique())}')
    print(f'Random Value is {data[col][np.random.randint(data.shape[0])]}')
    print(f'Random Value is {data[col][np.random.randint(data.shape[0])]}')
    print(f'Random Value is {data[col][np.random.randint(data.shape[0])]}')
    print('\n==================================')

In [12]:
show_column_details('Make')

for feature Make and it type is object
Number of Nulls is 0
Number of Unique Values is 90
Random Value is Chrysler
Random Value is Chevrolet
Random Value is Daewoo



there are many categorical unique values in "Make" Columns therefore we will handle it using label encoder 

In [13]:
show_column_details('Model')

for feature Model and it type is object
Number of Nulls is 0
Number of Unique Values is 857
Random Value is Verna
Random Value is Elantra
Random Value is F3



In [14]:
show_column_details('CC')

for feature CC and it type is object
Number of Nulls is 18678
Number of Unique Values is 107
Random Value is nan
Random Value is 1200 CC
Random Value is 4000 CC



In [15]:
data['CC'].unique()

array([nan, '1600 CC', '1400 CC', '2000 CC', '1500 CC', '1300 CC',
       '1200 CC', '1000 CC', '1100 CC', '2500 CC', '3000 CC', '4400 CC',
       '4700 CC', '2002 CC', '1800 CC', '900 CC', '4000 CC', '950 CC',
       '3700 CC', '3500 CC', '800 CC', '2300 CC', '2700 CC', '2200 CC',
       '3800 CC', '1989 CC', '3750 CC', '1399 CC', '3400 CC', '4100 CC',
       '4800 CC', '3646 CC', '1250 CC', '2800 CC', '1398 CC', '4500 CC',
       '3200 CC', '5700 CC', '2600 CC', '3600 CC', '2900 CC', '6000 CC',
       '2400 CC', '5200 CC', '1599 CC', '1716 CC', '1116 CC', '1900 CC',
       '2019 CC', '1998 CC', '4200 CC', '1999 CC', '1550 CC', '1323 CC',
       '995 CC', '1598 CC', '1897 CC', '1489 CC', '3300 CC', '1992 CC',
       '5600 CC', '3190 CC', '850 CC', '1139 CC', '2127 CC', '1295 CC',
       '1700 CC', '5400 CC', '1490 CC', '970 CC', '5800 CC', '1050 CC',
       '1850 CC', '1572 CC', '2609 CC', '4570 CC', '1970 CC', '1150 CC',
       '5000 CC', '4600 CC', '1595 CC', '5900 CC', '2100 CC', '

In [16]:
data['CC'].fillna(0 , inplace=True)

In [17]:
data['CC'] = data['CC'].astype(str).apply(lambda x :x.replace('CC',""))

In [18]:
data['CC'] = data['CC'].astype(int)

In [19]:
data['Is_CC'] = data['CC'].apply(lambda x:True if x != 0  else False)

In [20]:
data['Is_CC'].value_counts()

True     19067
False    18678
Name: Is_CC, dtype: int64

In [21]:
show_column_details('Km')

for feature Km and it type is object
Number of Nulls is 0
Number of Unique Values is 1032
Random Value is 220,000 Km
Random Value is 0 Km
Random Value is 0 Km



In [22]:
data['Km'].value_counts()

0 Km          7772
200,000 Km    2485
150,000 Km    1400
100,000 Km     950
130,000 Km     816
              ... 
408,000 Km       1
21,500 Km        1
130,500 Km       1
57,310 Km        1
186,100 Km       1
Name: Km, Length: 1032, dtype: int64

In [23]:
def replaceList(reList,text):
    for re in reList:
        text = text.replace(re , '')
    return int(text)
    

In [24]:
data['Km'] = data['Km'].astype(str).apply(lambda x :replaceList(['Km',','],x))

In [25]:
show_column_details('Price')

for feature Price and it type is object
Number of Nulls is 0
Number of Unique Values is 768
Random Value is 275,000 EGP
Random Value is 88,000 EGP
Random Value is 155,000 EGP



In [26]:
data['Price']= data['Price'].astype(str).apply(lambda x :replaceList(['EGP',','],x))

In [27]:
data['Used since'] = data['Used since'].astype(int)

In [28]:
data['Price'] = data['Price'].astype(int)

In [29]:
show_column_details('Transmission')

for feature Transmission and it type is object
Number of Nulls is 0
Number of Unique Values is 4
Random Value is automatic
Random Value is automatic
Random Value is manual



In [30]:
data['Transmission'].value_counts()

manual       19403
automatic    18108
CVT            155
DSG             79
Name: Transmission, dtype: int64

In [31]:
show_column_details('City')

for feature City and it type is object
Number of Nulls is 0
Number of Unique Values is 101
Random Value is Cairo
Random Value is Damietta
Random Value is Alexandria



In [32]:
show_column_details('Color')

for feature Color and it type is object
Number of Nulls is 0
Number of Unique Values is 24
Random Value is White
Random Value is Beige
Random Value is Gray



In [33]:
show_column_details('Body Style')

for feature Body Style and it type is object
Number of Nulls is 16460
Number of Unique Values is 12
Random Value is nan
Random Value is nan
Random Value is Sedan



In [34]:
data['Body Style'].fillna('None',inplace=True)

In [35]:
data['Body Style'].value_counts()

None                  16460
Sedan                 14773
Hatchback              3103
4x4                    1422
SUV                     552
Van                     528
MiniVans                401
Pick up                 398
Coupe                    73
Microbus                 24
Medium truck jumbo       10
Mini Pick up              1
Name: Body Style, dtype: int64

In [36]:
show_column_details('Fuel')

for feature Fuel and it type is object
Number of Nulls is 0
Number of Unique Values is 5
Random Value is gas
Random Value is gas
Random Value is gas



In [37]:
data['Fuel'].value_counts()

gas            37109
diesel           281
natural gas      266
Hybrid            77
Electric          12
Name: Fuel, dtype: int64

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37745 entries, 0 to 37837
Data columns (total 52 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Make                        37745 non-null  object
 1   Model                       37745 non-null  object
 2   CC                          37745 non-null  int32 
 3   Used since                  37745 non-null  int32 
 4   Km                          37745 non-null  int64 
 5   Transmission                37745 non-null  object
 6   City                        37745 non-null  object
 7   Color                       37745 non-null  object
 8   Body Style                  37745 non-null  object
 9   Fuel                        37745 non-null  object
 10  Leather seats               37745 non-null  bool  
 11  Closing mirrors             37745 non-null  bool  
 12  Intelligent parking system  37745 non-null  bool  
 13  Sunroof                     37745 non-null  bo

### Handling Boolean Values 

In [41]:
bool_columns = [col for col in data.columns if data[col].dtype == bool] 
len(bool_columns)

41

In [42]:
for col in bool_columns:
    data[col] = data[col].apply(lambda x : 1 if x == True else 0)

In [43]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37745 entries, 0 to 37837
Data columns (total 52 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Make                        37745 non-null  object
 1   Model                       37745 non-null  object
 2   CC                          37745 non-null  int32 
 3   Used since                  37745 non-null  int32 
 4   Km                          37745 non-null  int64 
 5   Transmission                37745 non-null  object
 6   City                        37745 non-null  object
 7   Color                       37745 non-null  object
 8   Body Style                  37745 non-null  object
 9   Fuel                        37745 non-null  object
 10  Leather seats               37745 non-null  int64 
 11  Closing mirrors             37745 non-null  int64 
 12  Intelligent parking system  37745 non-null  int64 
 13  Sunroof                     37745 non-null  in

In [44]:
data.to_csv('_final_processed.csv',index=False)